snapshot/localmounter.go

```diff
@@ -47,6 +47,16 @@ func (lm *localMounter) Mount() (string, error) {
 		lm.release = release
 	}
 
+	// Windows mounts don't use the target parameter of mount.All, they activate in-place.
+	if len(lm.mounts) == 1 && lm.mounts[0].Type == "windows-layer" {
```

Comment:
Why is this being called on `len(lm.mounts) == 1` only?

Reply:
Similar to the code immediately below for bind-mounts, we can only output a single path pointing to that mount if we only have one input path.

*Unlike* the bind-mounts code-path below, this is not an optimisation, but a requirement due to lack of containerd/containerd#2366.

Reply:
If multiple mounts are passed in, we fall through to the regular `mount.All` code at the end of the function, and return the name of the empty directory we created to mount onto.

I'm really not clear what the expected behaviour of `mount.All` is for multiple layers, I assume it only makes sense on Linux when `Type` is "overlay".

Reply:
If windows mount only contain single item then we can just remove that part. If not then add a `TODO` that this only works in a limited limited inputs.

Reply:
I'm not wildly-familiar with Go. Are you suggesting something like this?
```go
if lm.mounts[0].Type == "windows-layer"{
  if len(lm.mounts) != 1 {
    return "", errors.New("TODO: windows-layer only rationally supports mounting a single layer")
  }
 if err := mount.Mount(lm.mounts[0], ""); err != nil {
 //... the rest of this change
}
```

i.e. we handle "windows-layer" completely here, and don't let it flow down into the default `mount.All` behaviour?

Reply:
If windows mounts can only contain single item then yes. And if they can contain multiple then just let `mount.All` handle it. Letting it pass through to linux implementation like it is now seems always wrong.

Reply:
I honestly don't _know_ if they can contain multiple. Well, they can, I'm just not sure what the meaning is. The `mount.All` API is generic code that just loops across all the mounts in the list, using the same target directory, and calling `mount.Mount` which is platform-specific.

That said, unless the mount type is `overlay`, I'm not sure what the Linux version does with multiple mounts anyway. I guess just mounts each over the top of the last, so only the last one is visible.

The problem is that [*right now*](https://github.com/containerd/containerd/pull/2366), _target dir_ is not used on Windows, so multiple Windows mounts will each mount into their own directory, rather than mounting on top of each other at _target dir_.

And I don't know what's expected to happen once https://github.com/containerd/containerd/pull/2366 is resolved, in this case. There's two code-paths there, one uses symlinks and clears out any existing symlink first. The other code-path uses [`SetVolumeMountPoint`](https://docs.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-setvolumemountpointw) which I suspect will error when the second volume is mounted to the same place.

I think that the concept of `mount.All` is faulty on Windows, because a mount on Windows *includes* its parent layers, while on Linux, `mount.All` is (hopefully) only ever passing a list of layers to overlay into one directory. So there's no rational use-case for `mount.All` on Windows.

Reply:
In the end, perhaps the best approach is to move `Mount()` out into the _\_windows.go_ and _\_unix.go_ files instead, as is already done for `Unmount()`?

Reply:
I might also, next time I'm in this code, go looking to see what actually can call this, in case it's provable that Windows will only ever pass a single mount through to this API anyway.

Reply:
> Letting it pass through to linux implementation like it is now seems always wrong.

I'm not sure what you mean by this. Is it that `bind` and `rbind` are Linux-specific? The final block of code should be generic, barring deficiencies in the containerd Windows implementation. And in the multiple-mount case, it can't do any worse than we're already doing, and with much less chance of corrupting a layer by trying to delete its directory instead of unmounting it.

Reply:
> I'm not sure what you mean by this. 

Like you said, "The problem is that right now, target dir is not used on Windows, " so we shouldn't reach the code that does that. Either error in the multi-item case or handle only the first item like in single-item case.

> In the end, perhaps the best approach is to move Mount() out into the _windows.go and _unix.go files instead, as is already done for Unmount()?

That's fine by me as well.

Reply:
Next time I'm in this codebase, I'll pull this commit out of this series, and rework it separately to pull `Mount` into the target-split files, so the Windows one can error on a multi-target call without disrupting the flow of the general case.

If https://github.com/containerd/containerd/pull/2366 is later resolved in a useful way for multiple mounts, `Mount` can be reunified after such a change is vendored/moduled in.